text/0000-inline-const.md

```diff
@@ -0,0 +1,276 @@
+let v = [const { Vec::new() }; 3];  // IDIOMATIC
+let v = const { [ Vec::new(); 3] }; // NOT IDIOMATIC
+```
+
+Whether to lint against the non idiomatic versions is an open question.
+Personally, I would like to lint against `&const {...}` but not `const { [expr;
+3] }`.
+
+Inline `const`s are allowed within `const` and `static` initializers, just as we
+currently allow nested `const` declarations. Whether to lint against inline
+`const` expressions inside a `const` or `static` is also an open question.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This excludes other uses of the `const` keyword in expressions and patterns.
+I'm not aware of any other proposals that would take advantage of this.
+
+This would also be the first use of type inference for const initializers. I'm
+not aware of any technical issues that would arise from this, but perhaps I'm
+overlooking something?
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The main alternative is the status quo. Maintaining it will likely result in
+promotion being used for more contexts. The lang-team decided to [explore this
+approach](https://github.com/rust-lang/rust/pull/70042#issuecomment-612221597)
+instead.
+
+It would also possible to separate out the parts of this RFC relating to patterns
+so that they can be decided upon seperately. I think they are similar enough
+that they are best considered as a unit, however.
+
+# Prior art
+[prior-art]: #prior-art
+
+I'm not aware of equivalents in other languages.
+
+AFAIK, this was [first proposed] by @scottmcm.
+
+[first proposed]: https://internals.rust-lang.org/t/quick-thought-const-blocks/7803/9
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## Naming
+
+I prefer the name inline `const`, since it signals that there is no difference
+between a named `const` and an inline one. @scottmcm prefers "`const` block",
+which is closer to the syntax and parallels the current terminology of `async`
+block and `unsafe` block. It also avoids any accidental conflation with the
+`#[inline]` attribute, which is unrelated.
+
+## Lints
+
+As mentioned in the reference-level specification, we need to decide whether we
+want to lint against certain types of inline `const` expressions.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+It would be possible to allow the syntax `const expr` for an inline `const` that
+consists of a single expression. This is analagous to the single expression
+variant of closures: `|| 42`. This is backwards compatible with the current proposal.
+
+This could allow us to deprecate the more esoteric classes of promotable
+expressions (e.g., `&(u32::MAX + u32::MAX)`) in favor of inline `const`
+expressions. This would have to be done at an edition boundary. We would only
+do promotion for aggregates, literals, constants and combinations thereof, and
+`#[rustc_promotable]` would be removed from the standard library.
```

Comment:
I don't think this attribute could be removed as the standard library would still be used by older editions.  

```rust
// must still compile on 2018 edition 
let _: &'static u32 = &u32::max_value();
```

Reply:
In that case the attributes can remain, they would just be ignored by the 2021 compiler.

See rust-lang/rust#71796 for an attempt to further reduce the number of implicitly promotable functions.